### PR TITLE
Add VSCode devcontainer configuration

### PR DESCRIPTION
**Pull request by idan**
This creates the a basic harness that permits users to develop with their local checkout inside the same docker container which powers the GitHub action. Does not yet contain docs and tests, but I did try it out with a clean checkout and was able to build images. 🎉 

See #208 for discussion about where this can go.

## Steps to use:

1. Install docker
2. Install the [Remote - Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) either via the extensions panel, or in your favorite terminal:
`code --install-extension ms-vscode-remote.remote-containers`
3. Opening the `zmk` folder in VS Code now displays a prompt to re-open inside a container:
image
4. Click `Reopen in Container` and the VS Code window will reload and do the necessary initial docker setup. This can take a little while.
5. Open a terminal in VS Code and issue the following commands to initialize your environment:

```bash
$ west init -l app
$ west update # and go make lunch, this is going to take a while
$ cd app
$ west build -b YourBoardName -- -DSHIELD=YourShieldName
```

That's it!

### CODE REVIEW

**Comment by CrossR** [comment]
Some potential to combine this with #192, and include the Docker ext as a recommended extension, though also raises the question of it the instructions provided there work for a Docker based workflow (making the compilation database, using it etc).

**Comment by ChaosinaCan** [comment]
I suspect that the compilation database part is still correct, but you'd need to change the compiler path to match wherever it ends up being in the container. Maybe we just add Docker to the list of OSes in all the docs pages?

**Comment by petejohanson** [comment]
> I suspect that the compilation database part is still correct, but you'd need to change the compiler path to match wherever it ends up being in the container. Maybe we just add Docker to the list of OSes in all the docs pages?

Yeah, if we can make this seemless, I definitely think it should be an "OS tab/option" in the dev setup docs.

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "ZMK Development",
+  "dockerFile": "Dockerfile",
```
Comment (petejohanson) [inline .devcontainer/devcontainer.json:3]: It isn't easily found in the docs, but you can just replace this with:

```suggestion
  "image": "zmkfirmware/zephyr-west-action-arm",
```

And remove the extraneous `Dockerfile`.

The other thing I had to do locally here, to make this work on my Fedora machine w/ SELinux enforcement enabled, was to *also* add a line:

```
"runArgs": ["--security-opt", "label=disable"],
```

Does it work for you locally if you also have that in this file?

**Review by petejohanson — CHANGES_REQUESTED** [review]
One tweak, and a question around SELinux settings.

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (idan) [reply .devcontainer/devcontainer.json]: Yup, fixed in 164cd97

**Comment by idan** [comment]
So, the tweak above worked. Onto new issues!

Git pull/push with SSH remotes doesn't work inside the container because the docker image lacks SSH. Fixable with `apt update; apt install ssh`. We should probably bake that into the image? Once that's there, it Just Works™, and the devcontainers extension properly forwards things to the host's ssh-agent.

The more annoying (but less critical) thing I've run into is that GPG commit signing doesn't work. Like SSH, GPG is not installed in the image (easily fixable like above). Once it's there, it still doesn't work — the `~/.gnupg` directory doesn't even exist:

image


Supposedly, this [should work as of v118 of the remote-containers extension](https://github.com/microsoft/vscode-remote-release/issues/72), but there's no documentation. However, [another issue](https://github.com/microsoft/vscode-remote-release/issues/3221#issuecomment-654724269) identifies a control to test against. When I try to gpg from inside the "try a sample" node container, I can see my keyring and a socket, and signing works fine:

image

My current hunch is that maybe the image needs to contain GPG already when it boots or the VSCode remote-containers extension doesn't try to set up forwarding.

**Comment by idan** [comment]
To be clear:

* SSH is critical, you can't git without it in many cases
* Commit signing is nice to have
* You can always commit from outside the container; this is all about quality of life at this point

**Comment by idan** [comment]
OK, it works!

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (idan) [reply .devcontainer/devcontainer.json:3]: Ended up returning to the dockerfile approach so we could use it to install SSH and GPG.

**Review by petejohanson — COMMENTED** [review]
One comment on the dockerfile, but certainly minor. I would love to get this in, so we can build on it, even if we aren't 100% happy with the documentation piece.

Thoughts @idan ?

**Comment by idan** [comment]
@petejohanson yes! I'll hack on the docs tonight and try to get it into a happy place, or move them to a separate branch so we can commit the docker bits and do the docs in a follow-onw

**Review by petejohanson — APPROVED** [review]
This is fabulous! Thanks sticking with this, and the docs updates.

**Pull request merged**